text/0000-tool-attributes.md

```diff
@@ -0,0 +1,218 @@
+If this fails, then it reports a macro not found error. The compiler *may*
+suggest mis-typed attributes (declared or built-in).
+
+
+## Proposed for immediate implementation
+
+There is an attribute path white list of the names of tools shipped with the Rust
+distribution. Any crate can use an attibute path starting with those names and
+the attribute will not trigger the custom attribute lint or require a macro
+feature gate.
+
+E.g., `#[rustdoc::foo]` will be permitted in stable Rust code; `#[rustdoc]` will
+still be treated as a custom attribute.
+
+The initial list of allowed prefixes is `rustc`, `rustdoc`, and `rls`. As tools
+are added to the distribution, they will be allowed as path prefixes in
+attributes. We expect to add `rustfmt` and `clippy` in the near future. Note
+that whether one of these names can be used does not depend on whether the
+relevant component is installed on the user's system; this is a simple,
+universal white list.
+
+Given the earlier rules on name resolution, these attributes would shadow any
+attribute macro with the same name. This is not problematic because a macro
+would have to be in a module starting with a tool name (e.g., `rustdoc::foo`),
+naming macros in such a way is currently unstable, and this can be worked around
+by using an import (`use`).
+
+Tool-scoped attributes should be preserved by the compiler for as long as
+possible through compilation. This allows tools which plug into the compiler
+(like Clippy) to observe these attributes on items during type checking, etc.
+
+
+### Forward and backward compatability
+
+Since custom attributes are feature gated and scoped attributes are part of the
+unstable macros 2.0 work, there is no backwards compatability issue.
+
+For tools who want to move to these newly stable attributes (e.g., from
+`rustfmt_skip` to `rustfmt::skip`) they will have to manage the change
+themselves.
+
+Although the mechanism for opt-in for the long-term solution is unspecified, the
+actual usage of tool attributes seems pretty clear. Therefore we can be reasonably
+confident that this proposal is forward-compatible in its syntax, etc.
+
+For the white-listed tools, will their names be implicitly imported in the long-
+term solution? One could imagine either leaving them implicit (similar to the
+libraries prelude) or using warning cycles or an epoch to move them to explicit
+opt-in.
+
+Another hazard is that if in the long-term solution, the compiler checks
+attribute paths for validity rather than only the prefix. For example, if
+Rustfmt only defines `skip`, then `#[rustfmt::foo]` would not give an error
+under this proposal, but would in the long term. Given that `#[rustfmt::foo]` is
+almost certainly a mistake today, it seems easy to migrate to a stricter system
+with a warning cycle.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The proposed scheme does not allow tools or macros to use custom top-level
```

Comment:
A `use`d macro should be possible to use without prefix, no? Tools not, but I guess if somebody really wanted, they could create a macro to hide it from the compiler.

Reply:
Yes, this RFC does not affect imported macros.